kpatch-build/create-diff-object.c

```diff
@@ -321,6 +321,93 @@ static int kpatch_warn_only_change(struct section *sec)
 	return 1;
 }
 
+/*
+ * Determine if a section has changed only due to might_sleep's
+ * embedding of the line number into an instruction operand.
+ *
+ * Example:
+ *
+ *   50f:   be f7 01 00 00          mov    $0x1f7,%esi
+ *   514:   48 c7 c7 00 00 00 00    mov    $0x0,%rdi
+ *                          517: R_X86_64_32S       .rodata.do_select.str1.8+0x98
+ *   51b:   e8 00 00 00 00          callq  520 <do_select+0x520>
+ *                          51c: R_X86_64_PC32      ___might_sleep-0x4
+ */
+static int kpatch_debug_atomic_sleep_change(struct section *sec)
+{
+	struct insn insn1, insn2;
+	unsigned long start1, start2, size, offset, length;
+	struct rela *rela;
+	int warnonly = 0, found;
+
+	if (sec->status != CHANGED ||
+	    is_rela_section(sec) ||
+	    !is_text_section(sec) ||
+	    sec->sh.sh_size != sec->twin->sh.sh_size ||
+	    !sec->rela ||
+	    sec->rela->status != SAME)
+		return 0;
+
+	start1 = (unsigned long)sec->twin->data->d_buf;
+	start2 = (unsigned long)sec->data->d_buf;
+	size = sec->sh.sh_size;
+	for (offset = 0; offset < size; offset += length) {
+		insn_init(&insn1, (void *)(start1 + offset), 1);
+		insn_init(&insn2, (void *)(start2 + offset), 1);
+		insn_get_length(&insn1);
+		insn_get_length(&insn2);
+		length = insn1.length;
+
+		if (!insn1.length || !insn2.length)
+			ERROR("can't decode instruction in section %s at offset 0x%lx",
+			      sec->name, offset);
+
+		if (insn1.length != insn2.length)
+			return 0;
+
+		if (!memcmp((void *)start1 + offset, (void *)start2 + offset,
+			    length))
+			continue;
+
+		/* verify it's a mov immediate to %esi */
+		insn_get_opcode(&insn1);
+		insn_get_opcode(&insn2);
+		if (insn1.opcode.value != 0xbe || insn2.opcode.value != 0xbe)
+			return 0;
+
+		/* verify the next instruction is mov $0x0,%rdi */
+		if (*(unsigned int *)(start1 + offset + length) != 0xc7c748)
+			return 0;
```

Comment:
Setting CONFIG_DEBUG_ATOMIC_SLEEP and looking through the compiled code at calls to ___might_sleep(), this pattern doesn't always hold.  Sometimes it sets %esi first, sometimes it sets %rdi first, sometimes there are instructions in-between.  So I would remove this restriction.

After doing that, I think the only difference between this function and kpatch_warn_only_change() is the string compares inside the rela loop.  So I think it would be worthwhile to combine the two functions into a single `kpatch_line_macro_change_only()` function.  It might be as simple as adding the "__might_sleep" check at the end of the loop.